### ISSUE TITLE
Make it possible to limit memory usage of processes

### ISSUE THREAD

**Issue by pcmoritz** [ray-project/ray#3055]
Currently it is possible that if the worker processes take up too much memory, the linux OOM killer will kill random processes, which looks like random failures.

We should investigate how to limit memory consumption of workers and actors (hopefully there is a way to impose a limit on the sum of the memory consumption of multiple processes) and notify the user if memory consumption of custom code goes over.

(I was just running into this with one of our users)

**Comment by mitar** [comment]
Yes, we would need this as well. We want to evaluate our AutoML system under various conditions. Like how well it works with various sizes of search space, what is the best pipeline it can find given time limit, but also what it can best find given memory limit. For example, being allowed to construct pipelines which take 20 GB of RAM can find different solutions than if you can use at most 1 GB of RAM.

To me it seems the best way to achieve this is to limit how much memory a worker can consume. There seems to be [resource module](https://docs.python.org/3/library/resource.html) to set this.

So what is tricky here is to decide how much should be this limit. Is this `total_limit / number_of_workers`? `total_limit / number_of_workers_on_a_node`? `(total_limit - plasma_limit / number_of_workers`?

Should all workers get an equal memory chunk or should we just make sure that all workers together on a node do not use more memory? Do we want to require memory limit per job (we could set memory limit before starting a job, and then remove if after).

I think setting per job could help us doing evaluation I mentioned above. And setting for all workers together (but not per worker) could help us with another situation, which is that on Kubernetes, if you limit resources on the container with hard limit, Kubernetes will kill your pod if it consumes more resources. So your pod has to make sure it does not use more resources, which is currently lacking in Ray how to set.

**Comment by mitar** [comment]
One solution I am now trying is setting `/proc/pid/oom_adj` to `15` for all Ray workers. This makes Ray workers the most likely to be killed by OOM. Because I run Ray workers inside one Docker container, I can then set memory limit (cgroup) on the whole Docker container, which then means that OOM will be invoked when the container gets over the limit, but killing workers first before anything else in the container.

Based on simple testing it seems it works well.

**Comment by robertnishihara** [comment]
Nice, maybe we should add a flag to have an option to do that automatically.

**Comment by mitar** [comment]
```python
            with open('/proc/{pid}/oom_score_adj'.format(pid=os.getpid()), 'w') as oom_adj_file:
                oom_adj_file.write('1000')
```

**Comment by mitar** [comment]
Unstale.

**Comment by mitar** [comment]
Unstale.

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
